sdk/common/src/main/scala/org/typelevel/otel4s/sdk/common/InstrumentationScope.scala

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2023 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s.sdk
+package common
+
+import cats.Hash
+import cats.Show
+import cats.syntax.show._
+
+/** Holds information about instrumentation scope.
+  *
+  * Instrumentation scope is a logical unit of the application code with which
+  * emitted telemetry is associated. The most common approach is to use the
+  * instrumentation library as the scope, however other scopes are also common,
+  * e.g. a module, a package, or a class may be chosen as the instrumentation
+  * scope.
+  *
+  * @see
+  *   [[https://opentelemetry.io/docs/specs/otel/glossary/#instrumentation-scope Instrumentation Scope spec]]
+  */
+sealed trait InstrumentationScope {
+
+  /** Returns the name of the instrumentation scope.
+    */
+  def name: String
+
+  /** Returns the version of the instrumentation scope, or None if not
+    * available.
+    */
+  def version: Option[String]
+
+  /** Returns the URL of the schema used by this instrumentation scope, or None
+    * if not available.
+    */
+  def schemaUrl: Option[String]
+
+  /** Returns the attributes of this instrumentation scope.
+    */
+  def attributes: Attributes
+
+  override final def hashCode(): Int =
+    Hash[InstrumentationScope].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: InstrumentationScope =>
+        Hash[InstrumentationScope].eqv(this, other)
+      case _ =>
+        false
+    }
+
+  override final def toString: String =
+    Show[InstrumentationScope].show(this)
+}
+
+object InstrumentationScope {
+
+  /** A builder of [[InstrumentationScope]].
+    */
+  sealed trait Builder {
+
+    /** Sets the version.
+      *
+      * @param version
+      *   the version to set
+      */
+    def setVersion(version: String): Builder
```

Comment:
I don't want to be bikeshedding too much, it's not a big deal. And I guess otel4s already has some established idioms. But typically I associate `set` with mutability and e.g. `withVersion` would be more appropriate for an immutable builder like this one.

Reply:
Valid point. We primarily use `with*` across the otel4s codebase.

Reply:
I've updated the code